cf/data/dask_regrid.py

```diff
@@ -0,0 +1,590 @@
+        # weights matrix accordingly
+        # ------------------------------------------------------------
+        if method in ("conservative", "conservative_1st"):
+            # 1) First-order conservative method:
+            #
+            #     w_ji = f_ji * As_i / Ad_j
+            #
+            # where f_ji is the fraction of source cell i that
+            # contributes to destination cell j, As_i is the area of
+            # source cell i, and Ad_j is the area of destination cell
+            # j.
+            #
+            # If source grid cell i is masked then w_ji are set to
+            # zero for all destination cells j, and the remaining
+            # non-zero w_ji are divided by D_j, the fraction of
+            # destination cell j that intersects with unmasked cells
+            # of the source grid.
+            #
+            #     D_j = 1 - w_i1j - ... - w_iNj
+            #
+            # where w_iXj is the unmasked weight for masked source
+            # cell i and destination cell j.
+            D = 1 - weights[:, src_mask].sum(axis=1, keepdims=True)
+
+            # Get rid of values that are approximately zero, or
+            # spuriously negative. These values of 'D' correspond to
+            # destination cells that overlap only masked source
+            # cells. These weights will be zeroed later on, so it's OK
+            # to set their value to 1, i.e. a nice non-zero value that
+            # will allow us to divide by 'D' in the next step.
+            D = np.where(D < min_weight, 1, D)
+
+            # Divide the weights by 'D'. Note that for destination
+            # cells which do not intersect any masked source grid
+            # cells, 'D' will now be 1.
+            w = weights / D
+
+            # Zero the weights associated with masked source grid
+            # cells
+            w[:, src_mask] = 0
+
+            # Mask out rows of the weights matrix which contain all
+            # zeros. Such a row corresponds to a destination grid cell
+            # that does not intersect any unmasked source grid cells.
+            #
+            # Note: Rows that correspond to masked destination grid
+            #       cells will have already been masked.
+            w = np.ma.where(
+                np.count_nonzero(w, axis=1, keepdims=True), w, np.ma.masked
+            )
+
+        elif method in ("linear", "bilinear", "nearest_dtos"):
+            # 2) Linear and nearest neighbour methods:
+            #
+            # Mask out any row j that contains at least one positive
+            # (i.e. greater than or equal to 'min_weight') w_ji that
+            # corresponds to a masked source grid cell i. Such a row
+            # corresponds to a destination grid cell that intersects
+            # at least one masked source grid cell.
+            if np.ma.isMA(weights):
+                where = np.ma.where
+            else:
+                where = np.where
```

Comment:
Is this (not just using `np.ma.where` in any case) for performance reasons (at least that is my guess)?

Reply:
Yep - `np.ma.where` is considerably slower.